pkg/util/audit/audit.go

```diff
@@ -0,0 +1,158 @@
+package audit
+
+import (
+	"encoding/json"
+	"fmt"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/ARO-RP/pkg/util/audit/schema"
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	auditMetadataCategory    = "auditCategory"
+	auditMetadataCreatedTime = "auditCreatedTime"
+	auditMetadataFullPayload = "auditFullPayload"
+	auditMetadataLogKind     = "logKind"
+	auditMetadataOperation   = "auditOperation"
+	auditMetadataResult      = "auditResult"
+	auditMetadataSource      = "auditSource"
+	auditSourceRP            = "aro-rp"
+
+	ifxAuditCloudVer = 1.0
+	ifxAuditLogKind  = "ifxaudit"
+	ifxAuditName     = "#Ifx.AuditSchema"
+	ifxAuditVersion  = 2.1
+
+	// ifxAuditFlags is a collection of values bit-packed into a 64-bit integer.
+	// These properties describe how the event should be processed by the pipelinei
+	// in an implementation-independent way.
+	ifxAuditFlags = 257
+
+	// ifxAuditPopSample is the effective percentage sample rate for this event at the
+	// time it was generated by a client (either the default value specified in
+	// the event definition, system default or a dynamic override from a cloud
+	// settings service).
+	// not used by Azure
+	ifxAuditPopSample = 0.0
+)
+
+var (
+	// epoch is an unique identifier associated with the current session of the
+	// telemetry library running on the platform. It must be stable during a
+	// session, and has no implied ordering across sessions.
+	epoch string
+
+	// sequence field used to track absolute order of uploaded events, per session.
+	// it is reset when the ARO component is restarted.
+	seqNum      uint64
+	seqNumMutex *sync.Mutex
+)
+
+func init() {
+	epoch = uuid.NewV4().String()
+	seqNum = 1
+	seqNumMutex = new(sync.Mutex)
+}
+
+// EmitRPLog is used by aro-rp to emit audit logs.
+// Caller is responsible for providing the logrus.Entry object that contains
+// the appropriate formatter, hooks, log level etc.
+// All audit logs will be emitted at the INFO level.
+func EmitRPLog(entry logrus.Entry, log *Log) error {
+	return emitLog(entry, log, auditSourceRP)
+}
+
+func emitLog(entry logrus.Entry, log *Log, source string) error {
+	payload := log.toPayload()
+
+	fields := logrus.Fields{}
+	fields[auditMetadataCategory] = payload.Category
+	fields[auditMetadataCreatedTime] = time.Now().UTC().Format(time.RFC3339)
+	fields[auditMetadataLogKind] = ifxAuditLogKind
+	fields[auditMetadataOperation] = payload.OperationName
+	fields[auditMetadataResult] = payload.Result.ResultType
+	fields[auditMetadataSource] = source
+
+	marshalled, err := marshal(payload)
+	if err != nil {
+		return err
+	}
+	fields[auditMetadataFullPayload] = marshalled
+
+	// log message can be anything here, since the important audit data are
+	// captured in the different log fields above.
+	entry.WithFields(fields).Info("see auditFullPayload field for full log data")
+	return nil
+}
+
+// Log captures information about the operation to be logged.
+type Log struct {
```

Comment:
Discussion/consideration points rather than hard guidance:

Fundamentally, my first interest/priority is: what does this library actually look like at point of use, and how do we minimise the impact of it?  Could you add a real life example or two to this PR to help clarify?

I have a secondary/corollary interest which is: how do we minimise re-inventing the wheel and to what extent can the existing logrus library help us?  Logrus is definitely not the last word, but we do already use it a lot so being similar is maybe less to think about?

Today, we pass base *logrus.Entry's into components, with fields already set that are considered immutable for that component.  Then when we do log.Print() or log.WithFields() internally logrus clones the entry for us and adds the dynamic stuff.

1) Rather than having EmitRPLog() (and presumably in the future EmitPortalLog() and others?) what if we just start with an appropriately configured base *logrus.Entry for each component, pass it in and have logrus clone it?  You might want a helper function that prepares this base *logrus.Entry.

2) I wonder if the intermediate Log struct defined here is necessary and how much it complicates usability?  What if we just expose a *logrus.Entry and little more at point of use, and we use a Hook on the underlying Entry's Logger to do things just in time before printing like add the auditMetadataFullPayload field?

Reply:
Yes, I have been thinking about this a lot from the caller/usage perspective. I believe we share very similar (if not identical) priority.

Using `cmd/aro/rp.go` as an example (not saying this is what we will log), I envision something like this:

```go
diff --git a/cmd/aro/rp.go b/cmd/aro/rp.go                                                                                                        
index 20729486..ddcaf013 100644                                                                                                                   
--- a/cmd/aro/rp.go                                                                                                                               
+++ b/cmd/aro/rp.go                                                                                                                               
@@ -8,6 +8,7 @@ import (                                                                                                                          
        "fmt"                                                                                                                                     
        "os"                                                                                                                                      
        "os/signal"                                                                                                                               
+       "strconv"                                                                                                                                 
        "syscall"                                                                                                                                 
                                                                                                                                                  
        "github.com/Azure/go-autorest/tracing"                                                                                                    
@@ -27,8 +28,11 @@ import (                                                                                                                       
        "github.com/Azure/ARO-RP/pkg/metrics/statsd"                                                                                              
        "github.com/Azure/ARO-RP/pkg/metrics/statsd/azure"                                                                                        
        "github.com/Azure/ARO-RP/pkg/metrics/statsd/k8s"                                                                                          
+       "github.com/Azure/ARO-RP/pkg/util/audit"                                                                                                  
+       "github.com/Azure/ARO-RP/pkg/util/audit/schema"                                                                                           
        "github.com/Azure/ARO-RP/pkg/util/deployment"
        "github.com/Azure/ARO-RP/pkg/util/encryption"
+       az "github.com/Azure/go-autorest/autorest/azure"
 )

 func rp(ctx context.Context, log *logrus.Entry) error {
@@ -60,6 +64,27 @@ func rp(ctx context.Context, log *logrus.Entry) error {
                }
        }
 
+       auditLog := &audit.Log{
+               AzureEnvironment: az.PublicCloud.Name,
+               CallerIdentities: []*schema.CallerIdentity{
+                       {
+                               CallerIdentityType:  schema.CallerIdentityTypePUID,
+                               CallerIdentityValue: strconv.Itoa(os.Getpid()),
+                       },
+               },
+               Category:      schema.CategoryAuthorization,
+               OperationName: "initializeAuthorizers",
+               Region:        _env.Location(),
+               TargetResources: []*schema.TargetResource{
+                       {
+                               TargetResourceType: "resource provider",
+                               TargetResourceName: "aro-rp",
+                       },
+               },
+       }
+       if err := audit.EmitRPLog(log, auditLog); err != nil {
+               return err
+       }
        err = _env.InitializeAuthorizers()
        if err != nil {
                return err
```

A few things worth noting:

* I am relying on the base `*logrus.Entry` object built by `log.GetLogger()`. This allows me to re-use the existing 
log-related configuration such as hooks, formatter, output mechanism (stdout, journald) etc.

*  The caller is responsible for collecting all the audit data such as target resources, caller identity etc. And 
depending on the nature of the event, some fields may be more relevant than others. 

E.g., `CallerIPAddress` will only be relevant for inbound request events, `TargetResources` will be more 
relevant to operations performed on cluster resources. 

Hence, there may not be enough common denominators to warrant another base `*logrus.Entry` object (based 
on the original `*logrus.Entry` base object) at the audit level.

> and we use a Hook on the underlying Entry's Logger to do things just in time before printing like add the auditMetadataFullPayload field

My impression is that the hook approach will work for fields that can be auto-hydrated, based on environment
specifics. In the example above, that will be the process ID. This will lead us to write code that can differentiate the 
different kinds of audit logs, and recognize which fields can be auto-populated.

If my understanding is correct, I am a bit on the fence about it. My hope is that at the moment, the caller is responsible
for assembling and providing all the audit logs data (minus metadata like log creation time, region etc.)

* Yes, my initial thought lean towards adding new exported functions like `EmitPortalLog()` as they become 
relevant. The rationale is that if I am adding audit logs in RP, there is only one exported function to call. Likewise, 
there is only one exported function for `aro-admin-portal`, `aro-proxy` etc. I don't need to know the details around 
making sure the `Source` field must be set to component (in this case, `aro-rp`) name.

LMKWYT.

Reply:
Per offline conversation, we will introduce a new constructor to return a `logrus.Entry` instance that has a payload hook added. The hook will know how to hydrate a IFxAudit payload object from the entry. Callers are responsible for providing and assembling this audit log data. Helper functions can be added in the next PR where we start adding the caller code.